Source/Scene/CesiumIon.js

```diff
@@ -0,0 +1,267 @@
+            // requests(since Image failures can not provide a status code)
+            if (!defined(error) || (error.statusCode !== 401 && !(error.target instanceof Image))) {
+                return when.resolve(false);
+            }
+
+            var ionData = that.ionData;
+
+            // We use a shared pending promise for all derived assets, since they share
+            // a common access_token.  If we're already requesting a new token for this
+            // asset, we wait on the same promise.
+            var pendingPromise = ionData._pendingPromise;
+            if (!defined(pendingPromise)) {
+                pendingPromise = CesiumIon._loadJson(endpointResource)
+                    .then(function(endpoint) {
+                        //Set the token for root resource that this (and other) resources were derived
+                        ionData._root.queryParameters.access_token = endpoint.accessToken;
+                        return endpoint.accessToken;
+                    })
+                    .always(function(accessToken) {
+                        // Pass or fail, we're done with this promise, the next failure should use a new one.
+                        ionData._pendingPromise = undefined;
+
+                        // We need this return because our old busted version of when
+                        // doesn't conform to spec of returning the result of the above `then`.
+                        return accessToken;
+                    });
+                ionData._pendingPromise = pendingPromise;
+            }
+
+            return pendingPromise.then(function(accessToken) {
+                // Set the new token for this resource
+                that.queryParameters.access_token = accessToken;
+                return true;
+            });
+        };
+    }
+
+    function createFactory(Type) {
+        return function(options) {
+            return new Type(options);
+        };
+    }
+
+    // These values are the unofficial list of supported external imagery
+    // assets in the Cesium ion beta. They are subject to change.
+    var ImageryProviderMapping = {
+        ARCGIS_MAPSERVER: createFactory(ArcGisMapServerImageryProvider),
+        BING: createFactory(BingMapsImageryProvider),
+        GOOGLE_EARTH: createFactory(GoogleEarthEnterpriseMapsProvider),
+        MAPBOX: createFactory(MapboxImageryProvider),
+        SINGLE_TILE: createFactory(SingleTileImageryProvider),
+        TMS: createTileMapServiceImageryProvider,
+        URL_TEMPLATE: createFactory(UrlTemplateImageryProvider),
+        WMS: createFactory(WebMapServiceImageryProvider),
+        WMTS: createFactory(WebMapTileServiceImageryProvider)
+    };
+
+    CesiumIonResource.prototype.createImageryProvider = function() {
+        var type = this.ionData.endpoint.type;
+        if (type === 'IMAGERY') {
+            return createTileMapServiceImageryProvider({ url: this });
```

Comment:
It always feels weird to me to pass `this` around in this way.  I would do something like `createImageryProvider(resource);` from `CesiumIon.createImageryProvider` instead.

Reply:
There's nothing weird about it at all.  I don't want to move this code into `CesiumIon.createImageryProvider` because then the `CesiumIonResource` one won't work on it's own (which is the eventual plan once `CesiumIonResource` is promoted to a first class public object)